modules/internal/ChapelArray.chpl

```diff
@@ -1788,9 +1808,41 @@ module ChapelArray {
       _value.dsiRequestCapacity(i);
     }
 
-    /* Return the number of indices in this domain */
-    proc size return _value.dsiNumIndices;
-    /* Return the lowest index in this domain */
+    /* Return the number of indices in this domain.  For domains whose
+       ``intIdxType != int``, please refer to the note in
+       :proc:`ChapelRange.range.size` about changes to this routine's return type.
+    */
+    proc size {
+      use ChapelRange;
+      if (chpl_idxTypeSizeChange(idxType) && sizeReturnsInt == false) {
+        compilerWarning("'.size' queries for domains and arrays with 'idxType == " +
+                        idxType:string +
+                        "' are changing to return 'int' values rather than '"+
+                        idxType:string+"'\n"+
+                        "  (to opt into the change now, re-compile with '-ssizeReturnsInt=true'\n" +
+                        "  or switch to the new '.sizeAs(type t: integral)' method)");
+        return this.sizeAs(this.intIdxType);
+      } else {
+        return this.sizeAs(int);
+      }
+    }
+
+    /* Return the number of indices in this domain as the specified type */
+    proc sizeAs(type t: integral): t {
+      use HaltWrappers;
+      const size = _value.dsiNumIndices;
+      if (size > max(t)) {
+        var error = ".size query exceeds max(" + t:string + ")";
+        if isRectangularDom(this) {
+          error += " for: '" + this:string + "'";
+        }
+        HaltWrappers.boundsCheckHalt(error);
```

Comment:
Not that it really matters since it's just to help us categorize halt, but I wonder if this is more appropriate as a castCheckHalt or maybe it's called safeCastCheckHalt. 

That also makes me wonder if this should only be enabled with boundsChecks or safeCastChecks instead of always on.

Reply:
Do we need to do any checking for negative numbers here? Assuming no since size should always be >=0, but I'm wondering if some user defined distribution could use negative size as a sentinel for something.

Reply:
In chapelrange this is guarded by `boundsChecking`, so I feel more confident we should use only do this check with runtime checks enabled.

Reply:
Good catch on the missing bounds-check test.

I was considering this a bounds check halt in the sense that the value we're trying to store into an integer is out of bounds for that integer type.  But I agree that it's in the grey area between bounds checking and safe casting.  I'm inclined to leave it alone if you're OK with that.

Interesting question about negative numbers.  If an implementation did use a negative size as a sentinel, is that something we should preserve and pass through or complain about here in the generic code?  I don't feel like I know what's right, so am inclined to leave that to the first person to run into a problem with having the dsi routine return a negative number to wrestle with.

Reply:
I'm fine leaving it as a bounds check halt. 

I'm also ok leaving the question of negative numbers for later